sympy/matrices/common.py

```diff
@@ -2167,6 +2170,14 @@ def __matmul__(self, other):
     def __mod__(self, other):
         return self.applyfunc(lambda x: x % other)
 
+    def mul(self, exp, simplify=True):
+        oldmulsimp              = self._threadlcl.mulsimp
+        self._threadlcl.mulsimp = simplify
+        ret                     = self.__mul__(exp)
```

Comment:
I don't think we want a thread local variable.

What I was imagining is this:
1. Move some (not sure which parts) of the logic from `Matrix.__mul__` into `mul`.
2. Add the argument to control simplification to `mul`.
3. Have `__mul__` call `mul` with whatever we think the default options should be (stick to current behaviour first)
4. Possibly `mul` can just replace `_eval_matrix_mul` but as a *public* function.

Then Add a docstring to `mul` that explains the parameters with examples and explains how they relate to `__mul__`. Then add it to the matrices tutorial in doc/src/tutorial/matrices.rst. I'm not sure if `mul` needs to be explicitly listed somewhere in `doc/src/modules/matrices/matrices.rst` or if there should be a section there that explains blowup.

Then for pow/power do the same except except use explicit calls to `mul` internally.

Reply:
Another couple of power outages here. Something like this then? Note how many places I had to touch just to pass the flag through, for me this is far messier than the TLS method but if you wanna do it like this then I'll clean it up and add it to the other classes.

> Possibly mul can just replace _eval_matrix_mul but as a public function.

A general `mul` makes sense but alongside not replacing `_eval_matrix_mul` due to there being several of these including `scalar` and `elementwise` versions.

Reply:
This is roughly what I meant although I only expected it to be for `_eval_pow_by_recursion` rather then propagating through Jordan form etc. All the testing and timings above have been about blowup in `_eval_pow_by_recursion`. Computing the Jordan form is very different and I think requires separate analysis and possibly a different method of blowup control anyway.

I don't know what you mean by the other classes but are you sure it needs to be added to them?

Reply:
I guess the point as I see it is that if we add `mul(..., mulsimp=True)` then it becomes possible to consider using it in all sorts of places. So far we have considered using it in `_eval_pow_by_recursion` and there's a strong case for at least providing users with the option for that via pow.

What I don't think we should do is propagate a mulsimp flag everywhere. There will be some cases where it is always desirable to simplify, others where it isn't needed at all and others where post-simplification as better or equally good. The thing about `_eval_pow_by_recursion` is that in many cases intermediate simplification is a lot more effective than post-simplification. That won't be true everywhere though.

Reply:
Ok here you go, starting with the simplest possible version - `mul`, `pow` and `_eval_pow_by_recursion` with a pass-through flag, exactly as requested if I understood you correctly. Now to clarify what I meant about other classes and things before I actually do any work:
1. This is currently applied to DenseMatrix but SparseMatrix should have also and also the default implementation of MatrixArithmetic._eval_matrix_mul for other matrices no?
2. What about scalar or even element-wise multiplication?
3. What about `__matmul__`, `__rmul__`, `__rmatmul__`, `__div__`?